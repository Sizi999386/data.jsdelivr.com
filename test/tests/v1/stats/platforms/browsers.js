const { expect } = require('chai');
const { makeEndpointSnapshotTests, makeEndpointPaginationTests, makePaginatedEndpointAssertions, setupSnapshots } = require('../../../../utils');

describe('/v1/stats/platforms/browsers', () => {
	before(() => {
		setupSnapshots(__filename);
	});

	makeStatsPlatformsVersionTests();
});

function makeStatsPlatformsVersionTests () {
	makePaginatedEndpointAssertions('/v1/stats/platforms/platform%2019/browsers{?continent,country,period}', {}, [
		{
			params: { period: '2020-04' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.sumBy(response.body, 'prev.share')).to.be.closeTo(100, .1);
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020-04', continent: 'EU' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.sumBy(response.body, 'prev.share')).to.be.closeTo(100, .1);
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020-04', country: 'PL' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.sumBy(response.body, 'prev.share')).to.be.closeTo(100, .1);
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020-02' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020-02', continent: 'EU' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020-02', country: 'PL' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020', continent: 'EU' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2020', country: 'PL' },
			assert: (response) => {
				expect(_.sumBy(response.body, 'share')).to.be.closeTo(100, .1);
				expect(_.every(response.body, result => result.prev.share === null)).to.be.true;
				expect(response.body).to.have.lengthOf(2);
				expect(response).to.matchSnapshot();
			},
		},
		{
			params: { period: '2019' },
			assert: (response) => {
				expect(response.body).to.be.empty;
			},
		},
	]);

	makeEndpointPaginationTests('/v1/stats/platforms/platform%2019/browsers', { period: '2020-04' });

	makeEndpointSnapshotTests('/v1/stats/platforms/platform%2019/browsers{?continent,country,period}', {}, [
		{ continent: 'X' },
		{ country: 'X' },
		{ continent: 'EU', country: 'PL' },
		{ period: 'month' },
	], { status: 400 });
}
