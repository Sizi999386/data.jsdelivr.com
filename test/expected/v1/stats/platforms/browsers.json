{
	"/v1/stats/platforms/platform%2019/browsers?continent=EU&country=PL": {
		"date": "2010-12-31",
		"status": 400,
		"message": "Invalid parameter value: object contains a conflict between optional exclusive peers [continent, country]."
	},
	"/v1/stats/platforms/platform%2019/browsers?continent=EU&period=2020": [
		{
			"browser": "browser 19",
			"share": 52.3,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.7,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?continent=EU&period=2020-02": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?continent=EU&period=2020-04": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": 51.19
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": 48.81
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?continent=X": {
		"date": "2010-12-31",
		"status": 400,
		"message": "Invalid parameter value: `continent` must be a valid continent code in uppercase."
	},
	"/v1/stats/platforms/platform%2019/browsers?country=PL&period=2020": [
		{
			"browser": "browser 19",
			"share": 52.3,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.7,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?country=PL&period=2020-02": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?country=PL&period=2020-04": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": 51.19
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": 48.81
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?country=X": {
		"date": "2010-12-31",
		"status": 400,
		"message": "Invalid parameter value: `country` must be a valid ISO 3166-1 alpha-2 country code in uppercase."
	},
	"/v1/stats/platforms/platform%2019/browsers?period=2020": [
		{
			"browser": "browser 19",
			"share": 52.3,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.7,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?period=2020-02": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": null
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": null
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?period=2020-04": [
		{
			"browser": "browser 19",
			"share": 52.34,
			"prev": {
				"share": 51.19
			}
		},
		{
			"browser": "browser 18",
			"share": 47.66,
			"prev": {
				"share": 48.81
			}
		}
	],
	"/v1/stats/platforms/platform%2019/browsers?period=month": {
		"date": "2010-12-31",
		"status": 400,
		"message": "Invalid parameter value: `period` must be one of [s-month, s-year] or a valid date in one of the following ISO formats [YYYY, YYYY-MM]."
	}
}
